### ISSUE TITLE
nixos/doc: Source links are broken, pointing at "release-19.09"

### ISSUE THREAD

**Issue by gloaming** [NixOS/nixpkgs#66917]
**Describe the bug**
"Declared by" source links are broken. in the NixOS manual's `Appendix A. Configuration Options` 

**To Reproduce**
Run `nixos-help`
Click `next` to go to Appendix A
Click on a source link

**Expected behavior**
I should land at the appropriate link for my Nixpkgs version, e.g. https://github.com/NixOS/nixpkgs/blob/8746c77a383/nixos/modules/config/appstream.nix

**Actual behaviour**
I land at a broken link for "release-19.09" e.g.
https://github.com/NixOS/nixpkgs/blob/release-19.09/nixos/modules/config/appstream.nix

**Metadata**
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.2.8, NixOS, 19.09pre188690.8746c77a383 (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2.2`
 - channels(craig): `""`
 - channels(root): `"nixos-19.09pre188690.8746c77a383"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by FRidh** [comment]
I suppose those links were never made with unstable in mind, only releases.

**Comment by matthewbauer** [comment]
Yeah we don't really differentiate between released and unreleased. Those links just come from the revision field:

https://github.com/NixOS/nixpkgs/blob/86f31274eb5e26cdb9f062ac7cbf06514ead1491/doc/doc-support/lib-function-locations.nix#L60

**Comment by gloaming** [comment]
It appears that it _is_ supposed to be the git revision, but it's falling back:

https://github.com/NixOS/nixpkgs/blob/86f31274eb5e26cdb9f062ac7cbf06514ead1491/doc/doc-support/lib-function-locations.nix#L3

```
nix-repl> lib.trivial.revisionWithDefault null
"3f6d0a981a56a55c47cdbdc5f44ddb0372d00909"
```

https://github.com/NixOS/nixpkgs/blob/86f31274eb5e26cdb9f062ac7cbf06514ead1491/lib/trivial.nix#L151-L160

```
$ cat /nix/var/nix/profiles/per-user/root/channels/nixos/.git-revision 
54c76638733ca5e85aefea3caefd3286e8b537c5
```

So why doesn't it work? :confused:

**Comment by ncfavier** [comment]
There are two different things here:

- The revision used for the links to *option declarations* in the NixOS manual (what this issue is about AFAICT) is currently determined by `lib.trivial.release`, which points to the *next* release on the unstable branches, hence the broken links. This is fixed by the introduction of `manualRevision` in https://github.com/NixOS/nixpkgs/pull/199363.
- The revision used for the links to the sources of *library functions* in the nixpkgs manual, which is determined by [nixpkgs/doc/doc-support/lib-function-locations.nix](https://github.com/NixOS/nixpkgs/blob/86f31274eb5e26cdb9f062ac7cbf06514ead1491/doc/doc-support/lib-function-locations.nix#L60). This doesn't seem too broken, except that it points to master even in the stable manual, so it should probably be made to use `manualRevision` as well.

**Comment by jtojnar** [comment]
The manual rendering has since been reimplemented. The new n-r-d toolchain supports linking to `master` but expects `local` as a revision to do that:

https://github.com/NixOS/nixpkgs/blob/40066ba0d8bb20a611c618fffa7ee0706c5de6d5/pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/options.py#L59-L62

but that does not appear to be set anywhere.

Nixpkgs manual uses `master` instead:

https://github.com/NixOS/nixpkgs/blob/87bc60cf2e7104c368d8a3c5a9302ad1b8d3d864/doc/default.nix#L137

And NixOS manual receives it from argument to `default.nix`:

https://github.com/NixOS/nixpkgs/blob/87bc60cf2e7104c368d8a3c5a9302ad1b8d3d864/nixos/doc/manual/default.nix#L125

Which is set in the documentation module to a hardcoded version:

https://github.com/NixOS/nixpkgs/blob/6ee6ce815a89ac0d1e105884f5a7223005e92b5e/nixos/modules/misc/documentation.nix#L57

**Comment by sternenseemann** [comment]
This is still an issue.

**Comment by ncfavier** [comment]
A fix can be extracted from https://github.com/NixOS/nixpkgs/pull/203994/commits/e9fe624505cb6525be18852c7c53aa3542ce18f3. This will require a corresponding addition to [this section](https://nixos.github.io/release-wiki/Branch-Off.html#on-the-release-branch) of the release wiki.
